### ISSUE TITLE
Linux build: unsupported cflag fvisibility passed to gcc

### ISSUE THREAD

**Issue by marscher** [omnia-md/conda-recipes#546]
As it happened in all recent PRs (eg: https://travis-ci.org/omnia-md/conda-recipes/jobs/139170394#L2575 or https://travis-ci.org/omnia-md/conda-recipes/jobs/138610361#L3793) I wonder if the holy build box has been updated?

Or are these flags dragged in by distutils/setuptools since Anacondas Python has been built with this (exotic) flag?

**Comment by marscher** [comment]
#541 #545

**Comment by rmcgibbo** [comment]
I don't think it's related to the build image, since I don't think that image has been updated and it doesn't autoupdate. But it could be related to a new version of any intervening package, or to the Python build itself.

**Comment by rmcgibbo** [comment]
I would think it should be `-fvisibility=hidden`, but maybe somewhere the `hidden` is being removed?

**Comment by marscher** [comment]
On 21.06.2016 15:50, Robert T. McGibbon wrote:
> I would think it should be |-fvisibility=hidden|, but maybe somewhere
> the |hidden| is being removed?
Thats the crucial question: where does this setting come from?

**Comment by jchodera** [comment]
Just to confirm: I haven't updated the omnia build box in four months:
https://hub.docker.com/r/jchodera/omnia-build-box/builds/

**Comment by jchodera** [comment]
I'm curious whether this addition of compiler arguments might have triggered some other arguments to be added...?
https://github.com/markovmodel/PyEMMA/commit/6adf0d87d35e39380f7aa4b3b9e60d91242d0b28

**Comment by mpharrigan** [comment]
holy build box defines some cflags

```
Holy build box activated
Prefix: /hbb_exe
CFLAGS: -g -O2 -fvisibility=hidden -I/hbb_exe/include 
LDFLAGS: -L/hbb_exe -static-libstdc++
STATICLIB_CFLAGS: -g -O2 -fvisibility=hidden -I/hbb_exe/include 
SHLIB_CFLAGS: -g -O2 -fvisibility=hidden -I/hbb_exe/include 
SHLIB_LDFLAGS: -L/hbb_exe -static-libstdc++
```

and then the compilation that fails (for msmbuilder)

```
gcc -pthread -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -g -O2 -fvisibility -fPIC -I/anaconda/envs/_build/lib/python3.5/site-packages/numpy/core/include -I/anaconda/envs/_build/include/python3.5m -c msmbuilder/example_datasets/_muller.c -o build/temp.linux-x86_64-3.5/msmbuilder/example_datasets/_muller.o
```

I'm trying to replicate this locally with `conda build` but it doesn't respect the cflags at all

**Comment by mpharrigan** [comment]
Although the docs 

conda.pydata.org/docs/building/environment-vars.html

Say that environment variables are not inherited, and CFLAGS isn't set on linux

**Comment by jchodera** [comment]
Weeeeird...

**Comment by mpharrigan** [comment]
I can reproduce with the docker image

**Comment by rmcgibbo** [comment]
I modifed a recipe's `build.sh` to `echo $CFLAGS` before running `setup.py`, and it prints...

```
-g -O2 -fvisibility
```

**Comment by rmcgibbo** [comment]
This has to be a conda-build bug.

**Comment by rmcgibbo** [comment]
```
$ export CFLAGS="-g -robert -mcgibbon -fvisibility=hidden"
$ conda-build ...
[snipped]
gcc -pthread -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -g -robert -mcgibbon -fvisibility -fPIC -DUSE_OPENMP -I/anaconda/envs/_build/lib/python3.5/site-packages/mdtraj/core/lib -I/anaconda/envs/_build/lib/python3.5/site-packages/numpy/core/include -Ipyemma/coordinates/clustering/include -I/anaconda/envs/_build/include/python3.5m -c pyemma/coordinates/clustering/src/regspatial.c -o build/temp.linux-x86_64-3.5/pyemma/coordinates/clustering/src/regspatial.o -std=c99 -g -O3 -pg -fopenmp
gcc: error: unrecognized command line option ‘-robert’
gcc: error: unrecognized command line option ‘-mcgibbon’
gcc: error: unrecognized command line option ‘-fvisibility’
```

**Comment by mpharrigan** [comment]
I agree. My investigative data points:

 - running python setup.py directly works
 - running `CFLAGS='' conda build ...` works

**Comment by mpharrigan** [comment]
That being said: it doesn't explain why I can't reproduce the problem outside of the docker container

Can we unset the cflags that are being set by holy build box?

**Comment by rmcgibbo** [comment]
Try setting `export CFLAGS=-hello=world` and running setup.py outside of the docker container, and you will reproduce it.

**Comment by rmcgibbo** [comment]
The bug is here: https://github.com/conda/conda-build/blob/master/conda_build/environ.py#L388-L389

**Comment by mpharrigan** [comment]
omg I couldn't reproduce because I didn't update my `conda-build ` even though I thought I did. At least now it makes sense

**Comment by rmcgibbo** [comment]
I'll file an issue.

**Comment by rmcgibbo** [comment]
lol why is all software terrible.

**Comment by peastman** [comment]
Because people who write software suck.

**Comment by jchodera** [comment]
Agile == fragile

**Comment by rmcgibbo** [comment]
Going forward, we could pin the version of conda-build that we use to avoid this kind of thing. I'm not sure if that would prevent the bulk of the issues, but it's an option.

**Comment by marscher** [comment]
I'm happy with pinning conda-build (until we discover the next bug in the pinned version)...

The fixed version is 1.20.1

Thank you guys for figuring this out!
